File: imagepicker/src/main/java/com/nguyenhoanglam/imagepicker/helper/ImageHelper.java
package com.nguyenhoanglam.imagepicker.helper;

import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.MediaStore;

import androidx.core.content.FileProvider;

import com.nguyenhoanglam.imagepicker.model.Image;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * Created by hoanglam on 7/31/16.
 */
public class ImageHelper {

    private static final String TAG = "ImageHelper";

    public static String getNameFromFilePath(String path) {
        if (path.contains(File.separator)) {
            return path.substring(path.lastIndexOf(File.separator) + 1);
        }
        return path;
    }

    public static void grantAppPermission(Context context, Intent intent, Uri fileUri) {
        List<ResolveInfo> resolvedIntentActivities = context.getPackageManager()
                .queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);

        for (ResolveInfo resolvedIntentInfo : resolvedIntentActivities) {
            String packageName = resolvedIntentInfo.activityInfo.packageName;
            context.grantUriPermission(packageName, fileUri,
                    Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION);
        }
    }

    public static void revokeAppPermission(Context context, Uri fileUri) {
        context.revokeUriPermission(fileUri,
                Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION);
    }

    public static List<Image> singleListFromPath(String path) {
        List<Image> images = new ArrayList<>();
        images.add(new Image(0, getNameFromFilePath(path), path));
        return images;
    }

    public static boolean isGifFormat(Image image) {
        String extension = image.getPath().substring(image.getPath().lastIndexOf(".") + 1, image.getPath().length());
        return extension.equalsIgnoreCase("gif");
    }


}

File: imagepicker/src/main/java/com/nguyenhoanglam/imagepicker/helper/ImageHelper.kt
/*
 * Copyright (C) 2023 Image Picker
 * Author: Nguyen Hoang Lam <hoanglamvn90@gmail.com>
 */

package com.nguyenhoanglam.imagepicker.helper

import android.net.Uri
import android.provider.MediaStore
import com.nguyenhoanglam.imagepicker.model.Folder
import com.nguyenhoanglam.imagepicker.model.Image

enum class SelectionState {
    NOT_SELECTED,
    SELECTED,
}

object ImageHelper {

    fun singleListFromImage(image: Image): ArrayList<Image> {
        val images = arrayListOf<Image>()
        images.add(image)
        return images
    }

    fun folderListFromImages(images: List<Image>): List<Folder> {
        val folderMap: MutableMap<Long, Folder> = LinkedHashMap()
        for (image in images) {
            val bucketId = image.bucketId
            val bucketName = image.bucketName
            var folder = folderMap[bucketId]
            if (folder == null) {
                folder = Folder(bucketId, bucketName)
                folderMap[bucketId] = folder
            }
            folder.images.add(image)
        }
        return ArrayList(folderMap.values)
    }

    fun filter(images: ArrayList<Image>?, bucketId: Long?): ArrayList<Image> {
        if (images == null) return arrayListOf()

        if (bucketId == null || bucketId == 0L) return ArrayList(images)

        return images.filter { it.bucketId == bucketId } as ArrayList<Image>
    }

    fun filterNot(images: ArrayList<Image>?, bucketId: Long?): ArrayList<Image> {
        if (images == null) return arrayListOf()

        if (bucketId == null || bucketId == 0L) return ArrayList(images)

        return images.filter { it.bucketId != bucketId } as ArrayList<Image>
    }

    fun filterExclude(
        images: ArrayList<Image>?,
        excludedImages: ArrayList<Image>?
    ): ArrayList<Image> {
        if (images.isNullOrEmpty()) return arrayListOf()

        if (excludedImages.isNullOrEmpty()) return ArrayList(images)

        return images.filter { findImageIndex(it, excludedImages) == -1 } as ArrayList<Image>

    }

    fun findImageIndex(image: Image, images: ArrayList<Image>): Int {
        for (i in images.indices) {
            if (images[i].uri == image.uri) {
                return i
            }
        }

        return -1
    }

    fun findImageIndexes(subImages: ArrayList<Image>, images: ArrayList<Image>): ArrayList<Int> {
        val indexes = arrayListOf<Int>()
        for (image in subImages) {
            for (i in images.indices) {
                if (images[i].uri == image.uri) {
                    indexes.add(i)
                    break
                }
            }
        }

        return indexes
    }

    fun getBucketSelectionState(
        images: ArrayList<Image>?,
        selectedImages: ArrayList<Image>?,
        bucketId: Long?,
    ): SelectionState {
        if (bucketId == 0L) {
            if (images?.isEmpty() == true) return SelectionState.NOT_SELECTED
            return if (selectedImages?.isEmpty() == true) SelectionState.NOT_SELECTED else SelectionState.SELECTED
        } else {
            if (images?.any { it.bucketId == bucketId } == false) return SelectionState.NOT_SELECTED
            return if (selectedImages?.any { it.bucketId == bucketId } == false) SelectionState.NOT_SELECTED else SelectionState.SELECTED
        }
    }


    fun isGifFormat(image: Image): Boolean {
        val fileName = image.name
        val extension = if (fileName.contains(".")) {
            fileName.substring(fileName.lastIndexOf(".") + 1, fileName.length)
        } else ""

        return extension.equals("gif", ignoreCase = true)
    }

    fun getImageCollectionUri(): Uri {
        return if (DeviceHelper.isMinSdk29) MediaStore.Images.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
        else MediaStore.Images.Media.EXTERNAL_CONTENT_URI
    }
}